File: out_go_v5/migrate/detect.go
package main

import (
    "errors"
    "fmt"
    "io"
    "log"
    "path/filepath"
    "sort"
    "strings"
    "unicode"

    "github.com/Masterminds/semver"
    "github.com/BurntSushi/toml"
)

type Language string

const (
    TypeScript Language = "TypeScript"
    JavaScript Language = "JavaScript"
    Python      Language = "Python"
    Go          Language = "Go"
    Rust        Language = "Rust"
    Java        Language = "Java"
    React       Language = "React"
    NextJs      Language = "Next.js"
    Vue         Language = "Vue"
    Svelte      Language = "Svelte"
)

var (
    Extensions = map[Language][]string{
        TypeScript: {"ts", "tsx"},
        JavaScript: {"js", "jsx", "mjs", "cjs"},
        Python: {"py"},
        Go: {"go"},
        Rust: {"rs"},
        Java: {"java"},
        React: {"jsx", "tsx"},
        NextJs: {"jsx", "tsx"},
        Vue: {"vue"},
        Svelte: {"svelte"},
    }
    Tools = map[Language]string{
        TypeScript: "npm",
        JavaScript: "npm",
        Python: "python",
        Go: "go",
        Rust: "cargo",
        Java: "javac",
        React: "npm",
        NextJs: "npm",
        Vue: "npm",
        Svelte: "npm",
    }
)

func (lang Language) Extensions() []string { return Extensions[lang] }

func (lang Language) TargetExtension() string {
    for ext := range Extensions[lang] {
        return strings.ToLower(string(ext))
    }
    return ""
}

func (lang Language) FromStr(s string) (Language, error) {
    s = strings.ToLower(toLowerCamel(s))
    for l, ext := range Extensions {
        if strings.HasPrefix(s, ext[0]...) {
            return l, nil
        }
    }
    return "", errors.New("unknown language")
}

func (lang Language) RequiredTool() string { return Tools[lang] }

func toLowerCamel(s string) string {
    var new string
    for _, r := range s {
        if unicode.IsUpper(r) {
            new += strings.ToLower(string(r))
        } else {
            new += string(r)
        }
    }
    return new
}

func ToolAvailable(tool string) bool {
    _ = semver.NewPreReleaseVersion(tool)
    stdout, _ := stdio.Open(tool)
    return stdout.Close() == nil
}

type Entry struct {
    Path string `toml:"path"`
    Info toml.FileInfo `toml:"info"`
}

type File struct {
    Path string   `toml:"path"`
    Size int64 `toml:"size"`
}

func main() {
    dir := "path_to_directory"
    lang := "TypeScript" // Replace with your language

    // Read TOML file with entries and paths
    file, err := os.Open("toml_file.toml")
    if err != nil {
        log.Fatal(err)
    }

    defer file.Close()

    var entries []Entry
    err = toml.Unmarshal(file, &entries)
    if err != nil {
        log.Fatal(err)
    }

    sort.Slice(entries, func(i, j int) bool { return entries[i].Path < entries[j].Path })

    // Find matching language for entries
    matchingLanguages := make(map[string]int)
    for _, e := range entries {
        for _, l := range []Language{TypeScript, JavaScript, Python} {
            for _, ext := range l.Extensions() {
                if strings.HasSuffix(e.Path, string(ext)) {
                    matchingLanguages[l]++
                }
            }
        }
    }

    // Print the language with the most matches
    max := 0
    for _, count := range matchingLanguages {
        if count > max {
            max = count
        }
    }

    var matching string
    for lang, count := range matchingLanguages {
        if count == max {
            matching += string(lang) + " "
        }
    }
    fmt.Println(matching)
}
```

To find the language of a file and recursively discover source files in a directory, use the following functions:

```go
func FileExtension(path string) string {
    ext := filepath.Ext(path)
    ext = strings.TrimPrefix(ext, ".")
    ext = strings.ToLower(ext)
    return ext
}

func FileLanguage(path string) Language {
    ext := FileExtension(path)
    for l, extensions := range Extensions {
        for _, ext := range extensions {
            if ext == ext {
                return l
            }
        }
    }
    return ""
}

func DiscoverSourceFiles(dir string, lang Language) []string {
    files := []string{}
    discoverRecursive(dir, lang, &files)
    sort.Strings(files)
    return files
}

func discoverRecursive(dir string, lang Language, files *[]string) {
    entries, err := filepath.Walk(dir, func(path string, info filepath.FileInfo, err error) error {
        if err != nil {
            return err
        }
        if !info.IsDir() {
            if matching := FileLanguage(path); matching == lang {
                *files = append(*files, path)
            }
        }
        return nil
    })
    if err != nil {
        log.Fatal(err)
    }
    return
}

File: core/src/migrate/detect.rs
use std::fmt;
use std::path::Path;

/// Supported programming languages for migration.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Language {
    TypeScript,
    JavaScript,
    Python,
    Go,
    Rust,
    Java,
    React,
    NextJs,
    Vue,
    Svelte,
    Cpp,
    Assembly,
}

impl Language {
    /// File extensions that belong to this language.
    pub fn extensions(&self) -> &'static [&'static str] {
        match self {
            Language::TypeScript => &["ts", "tsx"],
            Language::JavaScript => &["js", "jsx", "mjs", "cjs"],
            Language::Python => &["py"],
            Language::Go => &["go"],
            Language::Rust => &["rs"],
            Language::Java => &["java"],
            Language::React => &["jsx", "tsx"],
            Language::NextJs => &["jsx", "tsx"],
            Language::Vue => &["vue"],
            Language::Svelte => &["svelte"],
            Language::Cpp => &["cpp", "hpp", "cc", "h", "cxx"],
            Language::Assembly => &["asm", "s"],
        }
    }

    /// Target file extension when generating code in this language.
    pub fn target_extension(&self) -> &'static str {
        match self {
            Language::TypeScript => "ts",
            Language::JavaScript => "js",
            Language::Python => "py",
            Language::Go => "go",
            Language::Rust => "rs",
            Language::Java => "java",
            Language::React => "tsx",
            Language::NextJs => "tsx",
            Language::Vue => "vue",
            Language::Svelte => "svelte",
            Language::Cpp => "cpp",
            Language::Assembly => "asm",
        }
    }

    /// Parse a language string (case-insensitive).
    pub fn from_str_loose(s: &str) -> Option<Language> {
        match s.to_ascii_lowercase().as_str() {
            "ts" | "typescript" => Some(Language::TypeScript),
            "js" | "javascript" => Some(Language::JavaScript),
            "py" | "python" => Some(Language::Python),
            "go" | "golang" => Some(Language::Go),
            "rs" | "rust" => Some(Language::Rust),
            "java" => Some(Language::Java),
            "react" => Some(Language::React),
            "next" | "nextjs" => Some(Language::NextJs),
            "vue" => Some(Language::Vue),
            "svelte" => Some(Language::Svelte),
            "cpp" | "c++" => Some(Language::Cpp),
            "asm" | "assembly" => Some(Language::Assembly),
            _ => None,
        }
    }

    /// Name of the CLI tool needed for this language's ecosystem.
    pub fn required_tool(&self) -> &'static str {
        match self {
            Language::TypeScript | Language::JavaScript => "npm",
            Language::Python => "python",
            Language::Go => "go",
            Language::Rust => "cargo",
            Language::Java => "javac",
            Language::React | Language::NextJs | Language::Vue | Language::Svelte => "npm",
            Language::Cpp => "g++",
            Language::Assembly => "nasm",
        }
    }

    pub fn from_path(path: &Path) -> Language {
        detect_language(path).unwrap_or(Language::TypeScript)
    }
}

impl fmt::Display for Language {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Language::TypeScript => write!(f, "TypeScript"),
            Language::JavaScript => write!(f, "JavaScript"),
            Language::Python => write!(f, "Python"),
            Language::Go => write!(f, "Go"),
            Language::Rust => write!(f, "Rust"),
            Language::Java => write!(f, "Java"),
            Language::React => write!(f, "React"),
            Language::NextJs => write!(f, "Next.js"),
            Language::Vue => write!(f, "Vue"),
            Language::Svelte => write!(f, "Svelte"),
            Language::Cpp => write!(f, "C++"),
            Language::Assembly => write!(f, "Assembly"),
        }
    }
}

/// Detect the language of a file from its extension.
pub fn detect_language(path: &Path) -> Option<Language> {
    let ext = path.extension()?.to_str()?;
    let ext_lower = ext.to_ascii_lowercase();
    let all_languages = [
        Language::TypeScript,
        Language::JavaScript,
        Language::Python,
        Language::Go,
        Language::Rust,
        Language::Java,
        Language::React,
        Language::NextJs,
        Language::Vue,
        Language::Svelte,
        Language::Cpp,
        Language::Assembly,
    ];
    for lang in &all_languages {
        if lang.extensions().contains(&ext_lower.as_str()) {
            return Some(*lang);
        }
    }
    None
}

/// Directories to skip when discovering source files.
const SKIP_DIRS: &[&str] = &[
    "node_modules",
    "target",
    ".git",
    "__pycache__",
    ".venv",
    "venv",
    "dist",
    "build",
    ".next",
    ".idea",
    ".vscode",
    ".astra",
    ".forge",
    ".codex",
    "vendor",
    "bin",
    "obj",
];

pub fn discover_source_files(
    path: &Path,
    lang: Language,
) -> Vec<std::path::PathBuf> {
    let mut files = Vec::new();
    
    // Check if the path points to an existing file directly
    if path.is_file() || std::fs::metadata(path).map(|m| m.is_file()).unwrap_or(false) {
        if let Some(detected) = detect_language(path) {
            if detected == lang {
                files.push(path.to_path_buf());
            }
        }
    } else {
        discover_recursive(path, lang, &mut files);
    }
    files.sort();
    files
}


fn discover_recursive(
    dir: &Path,
    lang: Language,
    out: &mut Vec<std::path::PathBuf>,
) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };

    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            let name = path
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("");
            if SKIP_DIRS.contains(&name) {
                continue;
            }
            discover_recursive(&path, lang, out);
        } else if let Some(detected) = detect_language(&path) {
            if detected == lang {
                out.push(path);
            }
        }
    }
}

/// Check whether a required CLI tool is available on the system.
pub fn tool_available(tool: &str) -> bool {
    std::process::Command::new(tool)
        .arg("--version")
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .is_ok()
}
